sdk/servicebus/Azure.Messaging.ServiceBus/src/Primitives/ServiceBusRetryPolicy.cs

```diff
@@ -28,10 +28,11 @@ public abstract class ServiceBusRetryPolicy
     {
         private static readonly TimeSpan ServerBusyBaseSleepTime = TimeSpan.FromSeconds(10);
 
-        private readonly object serverBusyLock = new object();
-
-        // This is a volatile copy of IsServerBusy. IsServerBusy is synchronized with a lock, whereas encounteredServerBusy is kept volatile for performance reasons.
-        private volatile bool encounteredServerBusy;
+        /// <summary>
+        /// Represents a state flag that is used to make sure the server busy value can be observed with
+        /// reasonable fresh values without having to acquire a lock.
+        /// </summary>
+        private int serverBusyLock;
 
         /// <summary>
         /// Determines whether or not the server returned a busy error.
```

Comment:
Should `IsServerBusy` just do a `VolatileRead` of the flag variable rather than being managed as a separate state flag?   If I'm not mistaken, the read on L139 in the loop may potentially see a stale value as written.   _(I don't know that it is really impactful, but...)_

Reply:
Right now this only represents the state whether we entered the "critical" region and not the state of server busy. I can see if I can tweak things a bit

Reply:
Did you mean something like the following?

```csharp
        private bool IsServerBusy => Volatile.Read(ref serverBusy) == 1;

        private void SetServerBusy(string exceptionMessage, CancellationToken cancellationToken)
        {
            // multiple call to this method will not prolong the timer.
            if (Volatile.Read(ref serverBusy) == 1)
            {
                return;
            }

            ServerBusyExceptionMessage = string.IsNullOrWhiteSpace(exceptionMessage) ?
                Resources.DefaultServerBusyException : exceptionMessage;
            Volatile.Write(ref serverBusy, 1);
            _ = ScheduleResetServerBusy(cancellationToken);
        }

        private void ResetServerBusy()
        {
            if (Volatile.Read(ref serverBusy) == 0)
            {
                return;
            }

            ServerBusyExceptionMessage = Resources.DefaultServerBusyException;
            Volatile.Write(ref serverBusy, 0);
        }
``

Reply:
That's kind of what I was thinking, yup.

Reply:
Pushed a simplified version on the weekend that was inspired by how the `CancellationTokenSource` handles it states in a thread safe and lock free manner with the [underlying state field](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/Threading/CancellationTokenSource.cs#L38).